common/src/primitives/time.rs

```diff
@@ -50,12 +50,70 @@ pub fn set(now: Duration) -> Result<(), std::num::TryFromIntError> {
 }
 
 /// Either gets the current time or panics
-pub fn get_time() -> Duration {
+pub fn get_time() -> Time {
     match get_mocked_time() {
-        Some(mocked_time) => mocked_time,
-        None => SystemTime::now()
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .expect("Time went backwards"),
+        Some(mocked_time) => Time::from_duration_since_epoch(mocked_time),
+        None => Time::from_duration_since_epoch(
+            SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("Time went backwards"),
+        ),
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Time {
+    /// Time, stored as duration since SystemTime::UNIX_EPOCH
+    time: Duration,
+}
+
+impl Time {
+    pub const fn as_duration_since_epoch(&self) -> Duration {
+        self.time
+    }
+
+    pub const fn from_duration_since_epoch(duration: Duration) -> Self {
+        Self { time: duration }
+    }
+
+    pub const fn saturating_duration_add(&self, duration: Duration) -> Self {
+        Self {
+            time: self.time.saturating_add(duration),
+        }
+    }
+
+    pub const fn saturating_duration_sub(&self, t: Duration) -> Self {
+        Self {
+            time: self.time.saturating_sub(t),
+        }
+    }
+
+    pub const fn saturating_sub(&self, t: Self) -> Duration {
+        self.time.saturating_sub(t.time)
+    }
+}
+
+impl std::ops::Add<Duration> for Time {
+    type Output = Option<Self>;
+
+    fn add(self, other: Duration) -> Option<Self> {
+        self.time.checked_add(other).map(|time| Self { time })
+    }
+}
```

Comment:
Please don't take this the wrong way... but... the irony... comparing this to issue #1221... :smile:

Having said that: I do prefer enforcing checked math wherever possible... especially when it's global like this one.

I do understand there are many expects, but I'm OK with that as long as it's fully controlled. In fact, for uint, we're moving to further checked_add by default everywhere.

Reply:
As you say, but I'd argue that this comment and the mentioned issue have nothing in common.
Here we talking about whether to panic implicitly or explicitly. IMO for types like Duration implicit panics are OK and since with explicit ones we lose convenience (quite a lot of it IMO), I'd go with the former. But I can see your point too.
In the issue though, the question is whether we should panic in certain situations or not.